frontend/src/cljs/imas_seamap/views.cljs

```diff
@@ -100,21 +100,30 @@
     [:span.control.bp3-icon-large.bp3-icon-help.bp3-text-muted
      {:on-click #(re-frame/dispatch [:help-layer/open])}]]])
 
-(defn legend-display [{:keys [legend_url server_url layer_name style]}]
-  ;; Allow a custom url via the legend_url field, else construct a GetLegendGraphic call:
-  (let [legend-url (or legend_url
-                       (with-params server_url
-                         (merge
-                          {:REQUEST "GetLegendGraphic"
-                           :LAYER layer_name
-                           :FORMAT "image/png"
-                           :TRANSPARENT true
-                           :SERVICE "WMS"
-                           :VERSION "1.1.1"
-                           :LEGEND_OPTIONS "forceLabels:on"}
-                          (when style {:STYLE style}))))]
+(defn- vector-legend-rule [{:keys [title symbolizers] :as _rule}]
+  (let [color (-> symbolizers first :Polygon :fill)]
+    [:div.vector-legend-rule
+     [:div {:style {:background-color color}}]
+     [:div title]]))
+
+(defn- vector-legend [legend-info]
+  [:div
+   (map
+    (fn [{:keys [title] :as rule}]
+      ^{:key title}
+      [vector-legend-rule rule])
+    legend-info)])
+
+(defn legend-display [{:keys [legend_url] :as layer}]
+  (let [{:keys [has-info? loading? info]} @(re-frame/subscribe [:map.layer/legend layer])]
     [:div.legend-wrapper
-     [:img {:src legend-url}]]))
+     (cond
+       legend_url [:img {:src legend_url}] ; if we have a custom legend url, use that to display an image
+       has-info?  [vector-legend info]     ; else if we have the info for the layer then display it
+       loading?   [b/spinner]              ; else if we're loading show that
+       :else      (do                      ; else dispatch a request for the legend info
+                    (re-frame/dispatch [:map.layer/get-legend layer])
+                    [b/spinner]))]))
```

Comment:
I'm a bit fried and not instantly seeing the alternative, but this side-effect is making me feel icky.  (it's clever and I can see that it would work... I just want a cleaner way :))

Reply:
@olivergeorge mind weighing in here?  The problem is we want a component that displays a legend, but where previously the legend was just an img url, if we want vector legend support we have to make an ajax request.  Where do you think the best place to trigger that event is?  (I was musing in another part of the code that it might be better as a js component, rather than a bunch of reframe callback events.  That's still fairly side-effecty, so perhaps a wrapper that retrieves the contents, then a dumb rendering component?)

Reply:
Hmm, I guess the options are:
* in component (analog of image GET which happens as side effect of rendering)
* in handler (normal place for making requests with possible benefit of allowing data to be held in app-db for reuse) 

Generally I lean away from complex views which need to process data or embed business logic used to massage data before use.

So... it depends but there seem to be clear options and factors to help pick between them.

What am I missing?

Reply:
Resolved in Seamap chat 🙂